### DISCUSSION TITLE
I made a SHA256SUM-Checker-for-Windows, need help!

### DISCUSSION THREAD

**Discussion by gingshow** [community/community#21481] [Other Feature Feedback, Questions, & Ideas]
<p>This is how my [SHA256SUM-Checker-for-Windows] works.<br>
Step 1: The user needs to download the file he/she wants to verify and the SHA256SUM.txt.<br>
Step 2: The user needs to open SHA256SUM.txt manually.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/a1a94598ab3e87cb0cd51d749a9673e89599e9fe.jpeg" data-download-href="/uploads/short-url/n47xeviYx8v0BJd6Ripv5sYUvtY.jpeg?dl=1" title="2" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/35362003/181064953-3bd99912-b552-46f8-8bab-1600866be745.jpeg" alt="2" data-base62-sha1="n47xeviYx8v0BJd6Ripv5sYUvtY" width="690" height="102" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/a/a1a94598ab3e87cb0cd51d749a9673e89599e9fe_2_690x102.jpeg, https://aws1.discourse-cdn.com/github/optimized/2X/a/a1a94598ab3e87cb0cd51d749a9673e89599e9fe_2_1035x153.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/a/a1a94598ab3e87cb0cd51d749a9673e89599e9fe.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/a1a94598ab3e87cb0cd51d749a9673e89599e9fe_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">2</span><span class="informations">1342×200 164 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
Step 3: Then user needs to delete unnecessary part by themself.<br>
Step 4: Open hash.exe which is a command line application. Then the user needs to<br>
input the file name.<br>
Step 5: After that the program will generate the hash from the file.<br>
Step 6: Lastly, the user can copy the hash into the input field. The program will compare two hashes for you.<br>
But how can I make the program remove those unnecessary part for me? Thank you!<br>
For more screenshots please visit my repositories.<br>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/35362003/181064955-ed844465-2349-4ad2-a8ea-b95cbac3eb15.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows" target="_blank" rel="nofollow noopener">gingshow/SHA256SUM-Checker-for-Windows</a></h3>

<p>Contribute to gingshow/SHA256SUM-Checker-for-Windows development by creating an account on GitHub.</p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by gingshow** [comment]
<p>Step 2: The user needs to open SHA256SUM.txt manually.<br>
the above picture<br>
Step 3: Then user needs to delete unnecessary part by themself.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/28f8372b4d28fcebb85337a6ee82b6f56e40b692.jpeg" data-download-href="/uploads/short-url/5QqTGeu62w1sZKy9W10gdaOLYci.jpeg?dl=1" title="3" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/35362003/181064958-a61f5fec-f4f3-4c05-aed3-fcee31614f99.jpeg" alt="3" data-base62-sha1="5QqTGeu62w1sZKy9W10gdaOLYci" width="690" height="178" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/2/28f8372b4d28fcebb85337a6ee82b6f56e40b692_2_690x178.jpeg, https://aws1.discourse-cdn.com/github/original/2X/2/28f8372b4d28fcebb85337a6ee82b6f56e40b692.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/2/28f8372b4d28fcebb85337a6ee82b6f56e40b692.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/28f8372b4d28fcebb85337a6ee82b6f56e40b692_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">3</span><span class="informations">895×231 38.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
how can I make the program remove those unnecessary part for me? Thank you!<br>
For more screenshot please visit my repositories.
<p>Here is my code:<br>
</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/master/dist/hash.py" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/master/dist/hash.py" target="_blank" rel="nofollow noopener">gingshow/SHA256SUM-Checker-for-Windows/blob/master/dist/hash.py</a></h4>
<pre><code class="lang-py">import os
from termcolor import colored
import subprocess

if os.path.isfile('./sha256sum.txt'):
    print("file exists")
    file = open("sha256sum.txt", "r")

elif os.path.isfile('./SHA256SUMS.txt'):
    os.rename('SHA256SUMS.txt', 'sha256sum.txt')
    print("file exists")
    file = open("sha256sum.txt", "r")

elif os.path.isfile('./sha256.txt'):
    os.rename('sha256.txt', 'sha256sum.txt')
    print("file exists")
    file = open("sha256sum.txt", "r")

else:
    print("sha256sum.txt not found")
</code></pre>

  This file has been truncated. <a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/master/dist/hash.py" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="gingshow" data-post="1" data-topic="127990">
<div class="title">
<div class="quote-controls"></div>
 gingshow:</div>
<blockquote>
<p>But how can I make the program remove those unnecessary part for me?</p>
</blockquote>
</aside>
<p>Each line in the SHA-256 sums file has the hash, then spaces, then the filename. You can split that using a <a href="https://docs.python.org/3/library/re.html">regular expression</a> with two capturing groups. One thing to consider is how to find the right file(s) in that list. A common approach would be to expect the files in the list exist and have the same name on the local system, and to check them all (nonexistent files get an error message). The <a href="https://docs.python.org/3/library/pathlib.html"><code>pathlib</code></a> module might be helpful with handling file names.</p>
<p>A few other things I’m thinking of reading that code:</p>
<ul>
<li>When posting code, please use a code block (enclosed in triple backticks, ```). That preserves indentation, which is particularly critical in Python.</li>
<li>I strongly caution against renaming the checksum file. Users calling a check tool do not expect it to rename files.</li>
<li>You can very much simplify checking the different file names by using a loop over a list of possible filenames, and opening the first one that exists. However it might be better to accept any filename on the command line.</li>
<li>You don’t need powershell subprocesses, Python has a very helpful <a href="https://docs.python.org/3/library/hashlib.html"><code>hashlib</code></a> module. That has the advantage of working on any OS supported by Python. And it’ll be faster too, because it doesn’t need to start extra processes.</li>
</ul>

**Comment by gingshow** [comment]
<p>Thank you very much.</p>

**Comment by gingshow** [comment]
<p>Thank you for your suggestion!<br>
I made a new one.<br>
Here is the source code.<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/acd4ed46cb6fa078fba2a87dd8f1e616520a46e4/dist/hash.py" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/acd4ed46cb6fa078fba2a87dd8f1e616520a46e4/dist/hash.py" target="_blank" rel="nofollow noopener">gingshow/SHA256SUM-Checker-for-Windows/blob/acd4ed46cb6fa078fba2a87dd8f1e616520a46e4/dist/hash.py</a></h4>
<pre><code class="lang-py">import os
from termcolor import colored
import subprocess

hash1=""
hash2=""

if os.path.isfile('./sha256sum.txt'):
    print("sha256sum.txt exists")
    file1=open("sha256sum.txt","r")
else:
    print("sha256sum.txt does not exists,please make sure your file name is correct")
    input("Press enter to exit")
    exit(0)

if os.path.isfile('./sha256sum2.txt'):
    file2 = open("sha256sum2.txt", "w")
    file3 = open("sha256sum3.txt", "w")
else:
    print("sha256sum2.txt not exists.Creating it")
</code></pre>

  This file has been truncated. <a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/acd4ed46cb6fa078fba2a87dd8f1e616520a46e4/dist/hash.py" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
</p>

**Comment by airtower-luna** [comment]
<p>I’m sorry to say, but that won’t work at all: The code now checks if the hash entered by the user is the same as the last hash in the <code>sha256sum.txt</code> file, the hash of the actual file is ignored completely. And the two additional files are not needed at all, you should use variables to store data you will need later. And the point of reading the checksum file is that you don’t need to ask the user for a hash.</p>
<p>What I meant is to use a regular expression (using Python module <code>re</code>) to split each line in the <code>sha256sum.txt</code> file in its parts: The checksum and the filename. Something like this:</p>
<pre><code class="lang-auto">import re
r = re.compile(r'(^[0-9A-Fa-f]+)\s+(\S.*)$')
with open('sha256sum.txt') as fh:
    for line in fh:
        m = r.match(line)
        if m:
            checksum = m.group(1)
            filename = m.group(2)
            # compare here
</code></pre>
<p>Then for each <code>filename</code> check if the <code>checksum</code> matches the actual content. No need to ask the user for anything. If there’s anything unclear, feel free to ask! :slightly_smiling_face:</p>
<p>I couldn’t resist completing the example above myself, if you’d like to see the result look here:<br>
</p><aside class="onebox githubgist">
  <header class="source">
      <a href="https://gist.github.com/airtower-luna/a5df5d6143c8e9ffe7eb5deb5797a0e0" target="_blank" rel="noopener">gist.github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://gist.github.com/airtower-luna/a5df5d6143c8e9ffe7eb5deb5797a0e0" target="_blank" rel="noopener">https://gist.github.com/airtower-luna/a5df5d6143c8e9ffe7eb5deb5797a0e0</a></h4>
<h5>sha256check.py</h5>
<pre><code class="Python">#!/usr/bin/python3
# Example of how to verify an SHA-256 checksum file using Python.
# Usage: python sha256check.py sha256sum.txt
import hashlib
import re
import sys

# This regular expression matches a line containing a hexadecimal
# hash, spaces, and a filename. Parentheses create capturing groups.
r = re.compile(r'(^[0-9A-Fa-f]+)\s+(\S.*)$')</code></pre>
This file has been truncated. <a href="https://gist.github.com/airtower-luna/a5df5d6143c8e9ffe7eb5deb5797a0e0" target="_blank" rel="noopener">show original</a>

<p>
</p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<br>
It’s pure Python, so it’ll work on any OS/platform supported by Python.

**Comment by gingshow** [comment]
<p>Hi airtower-luna when I executed your code. It got an error. Why is this happened?</p>
<p>line 30, in <br>
with open(sys.argv[1]) as fh:<br>
IndexError: list index out of range</p>

**Comment by gingshow** [comment]
<p>I did some edited on my code according to your example code. Now my program can read the checksum file without asking the user for a hash.<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/e780f788ea56aad76e7213fa71dc96cb58ce033e/dist/hash.py" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/e780f788ea56aad76e7213fa71dc96cb58ce033e/dist/hash.py" target="_blank" rel="nofollow noopener">gingshow/SHA256SUM-Checker-for-Windows/blob/e780f788ea56aad76e7213fa71dc96cb58ce033e/dist/hash.py</a></h4>
<pre><code class="lang-py">import os
import termcolor
import subprocess
import hashlib

hash1=""
hash2=""

if os.path.isfile('./sha256sum.txt'):
    print("sha256sum.txt exists")
else:
    print("sha256sum.txt does not exists,please make sure your file name is correct")
    input("Press enter to exit")
    exit(0)

if os.path.isfile('./sha256sum2.txt'):
    file2 = open("sha256sum2.txt", "w")
    file3 = open("sha256sum3.txt", "w")
else:
    print("sha256sum2.txt not exists.Creating it")
</code></pre>

  This file has been truncated. <a href="https://github.com/gingshow/SHA256SUM-Checker-for-Windows/blob/e780f788ea56aad76e7213fa71dc96cb58ce033e/dist/hash.py" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="gingshow" data-post="8" data-topic="127990">
<div class="title">
<div class="quote-controls"></div>
 gingshow:</div>
<blockquote>
<p>It got an error. Why is this happened?</p>
<p>line 30, in<br>
with open(sys.argv[1]) as fh:<br>
IndexError: list index out of range</p>
</blockquote>
</aside>
<p>My script expects the name of the checksum file on the command line, instead of requiring it to be named <code>sha256sum.txt</code>. <a href="https://docs.python.org/3/library/sys.html#sys.argv"><code>sys.argv</code></a> (link goes to official Python documentation) is the list of command line arguments, starting with the name of the script itself. If there is no command line argument the list has no element <code>1</code>, so the access fails.</p>
<p>You could replace the <code>sys.argv[1]</code> with <code>'sha256sum.txt'</code> to use the same fixed name as in your script, but I prefer the flexibility. People can use different names for their checksum files after all. In a script that’s more than an example I’d check if there is an argument and write an error message saying how you should call the script if not, maybe like this:</p>
<pre><code class="lang-python">if len(sys.argv) &lt; 2:
    print('Missing argument, you need to provide a checksum file!', file=sys.stderr)
    print(f'Usage: {sys.argv[0]} CHECKSUM_FILE', file=sys.stderr)
    sys.exit(1)

with open(sys.argv[1]) as fh:
    # ... (as before)
</code></pre>
<p>There are a lot more possible error conditions, e.g the file might not exist or not have the right format.</p>
<p>As for the latest version of your script, the biggest issue is that it creates two new files (<code>sha256sum2.txt</code> and <code>sha256sum3.txt</code>) in the current directory. You don’t need those files, just keep the hashes in variables and compare them directly.</p>
<p>And in situations where you really need a file to store intermediate data, use a temporary file. Python has the <a href="https://docs.python.org/3/library/tempfile.html"><code>tempfile</code> module</a> to make that easy, including automatically deleting them on <code>close()</code>.</p>

**Comment by gingshow** [comment]
<p>Hey airtower-luna! Thank you very much.<br>
This is my newest version of my script.<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/gingshow/SHA256-Checker/blob/8ffdfd5da76ecfe88e8537f4c305b1175300329a/dist/hash.py" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/gingshow/SHA256-Checker/blob/8ffdfd5da76ecfe88e8537f4c305b1175300329a/dist/hash.py" target="_blank" rel="nofollow noopener">gingshow/SHA256-Checker/blob/8ffdfd5da76ecfe88e8537f4c305b1175300329a/dist/hash.py</a></h4>
<pre><code class="lang-py">import os
import hashlib

hash1=""
hash2=""

if os.path.isfile('./sha256sum.txt'):
    print("sha256sum.txt exists")
    file1 = open("sha256sum.txt", "r")
else:
    print("sha256sum.txt does not exists,please make sure your file name is correct")
    input("Press enter to exit")
    exit(0)

#Display list of files inside directory
print("Display list of files:")
for x in os.listdir('.'):
    print(x)

filename = input("Please enter the file name: ")
</code></pre>

  This file has been truncated. <a href="https://github.com/gingshow/SHA256-Checker/blob/8ffdfd5da76ecfe88e8537f4c305b1175300329a/dist/hash.py" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
</p>

**Comment by airtower-luna** [comment]
<p>Nice, those extra files are gone now. :slightly_smiling_face:</p>
<p>From a user point of view I’d still wonder why you need to ask for a filename when all the names are right there in the checksum file.</p>
<p>From my technical point of view I’m wondering: Why aren’t you using regular expressions to read the checksum file? :cat:</p>

**Comment by gingshow** [comment]
<p>Thank you :grinning:</p>
<p>Q: From a user point of view I’d still wonder why you need to ask for a filename when all the names are right there in the checksum file.<br>
A: The purpose of asking filename is I want the user have a choice to choose the file they want to check rather than the program get the check file for the user. :grin:</p>
<p>Q: From my technical point of view I’m wondering: Why aren’t you using regular expressions to read the checksum file?<br>
A: First,I haven’t learn how to use regular expressions. I will learn it later. :face_with_hand_over_mouth:<br>
Therefore, I choose another way, by reading the checksum file line by line, then the program can get the filename input from the user.<br>
After this the program will delete the filename, whitespaces and * from the checksum file. Then the program store those information into the variable without editing the originally checksum file. Finally, the program will check and compare the hash for the user.</p>
<p>This is my new SHA256-Checker repository location. Thank you!<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/gingshow/SHA256-Checker" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/35362003/181064963-5ecf7e7f-09f6-4c60-895b-b00aecc2ffda.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/gingshow/SHA256-Checker" target="_blank" rel="nofollow noopener">gingshow/SHA256-Checker</a></h3>

<p>A simply SHA256-Checker written in Python. Contribute to gingshow/SHA256-Checker development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="gingshow" data-post="13" data-topic="127990">
<div class="title">
<div class="quote-controls"></div>
 gingshow:</div>
<blockquote>
<p>First,I haven’t learn how to use regular expressions. I will learn it later. :face_with_hand_over_mouth:</p>
</blockquote>
</aside>
<p>Valid! :smile: I’d still recommend learning about them soon, they’re extremely useful for a lot of common programming problems, and this would be a good opportunity.</p>
<p>And I still think it’d be much more user-friendly to just verify all files listed in the checksum file than asking the user to type one of them. :wink:</p>
